pkg/urbit/vere/saga.c

```diff
@@ -424,17 +424,83 @@ u3_saga_open(const c3_path* const pax_u, u3_meta* const met_u)
   }
 
   try_list(log_u->epo_u.lis_u = c3_list_init(), goto free_dir_entries);
-  u3_epoc *poc_u, *pre_u;
+  u3_epoc *poc_u;
   c3_w* lif_w = &met_u->lif_w;
+  c3_d  nex_d = 0;
+  c3_l  mex_l = 0;
   for ( size_t idx_i = 0; idx_i < ent_i; idx_i++ ) {
     c3_path_push(log_u->pax_u, ent_c[idx_i]);
     try_epoc(poc_u = u3_epoc_open(log_u->pax_u, lif_w), goto free_dir_entries);
     c3_path_pop(log_u->pax_u);
+
+    if ( NULL != san_i ) {
+      //  check epoch's event range for contiguousness with earlier epochs.
+      //  first epoch may start at arbitrary event nr, so we don't check it.
+      //
+      c3_d fir_d = u3_epoc_first_commit(poc_u);
+      if ( (0 != nex_d) && (nex_d != fir_d) ) {
+        u3l_log("sane: epoch snapshot not contiguous: "
+                "expected %" PRIu64 ", got %" PRIu64 ", in %s\n",
+                nex_d, fir_d, ent_c[idx_i]);
+        *san_i = 1;
+        goto next;
+      }
+      nex_d = u3_epoc_last_commit(poc_u) + 1;
+
+      //  load the snapshot so we can validate its mug
+      //
+      if ( 0 != mex_l ) {
+        //NOTE  careful, will clobber all existing noun references!
+        c3_path_push(log_u->pax_u, ent_c[idx_i]);
+        u3m_init();
+        u3e_load(c3_path_str(log_u->pax_u));
+        u3m_pave(c3n);
+        c3_path_pop(log_u->pax_u);
+
+        if ( mex_l != u3r_mug(u3A->roc) ) {
+          u3l_log("sane: epoch snapshot not contiguous, mug mismatch in %s\n",
+                  ent_c[idx_i]);
+          *san_i = 2;
+          goto next;
+        }
+      }
+
+      //  read the next mug from the last event in the epoch
+      //
+      if ( !u3_epoc_iter_open(poc_u, u3_epoc_last_commit(poc_u)) ) {
+        u3l_log("sane: failed to open epoch event log\n");
+        *san_i = 3;
+        goto next;
+      }
+      c3_y*  dat_y;
+      size_t dat_i;
+      if ( !u3_epoc_iter_step(poc_u, &dat_y, &dat_i) ) {
+        u3l_log("sane: failed to read last event\n");
+        *san_i = 4;
+        u3_epoc_iter_close(poc_u);
+        goto next;
+      }
+      u3_epoc_iter_close(poc_u);
+      if ( dat_i < 4 ) {
```

Comment:
Let's replace `4` with `sizeof(mex_l)`:
```c
if ( dat_i < sizeof(mex_l) ) {
```

Reply:
But a mug is defined as always being four bytes? And we're gonna be reading out four bytes. The fact that we're storing it in a variable that has a matching size seems more incidental than something to rely on, if that makes sense. Using `sizeof` feels like it inverts the relationship.

Reply:
I see what you mean. Would it feel different if we used `sizeof(c3_l)`? I know `4` is used for "size of mug" elsewhere in the codebase, but I generally have an aversion to raw literals.

Reply:
It would not feel different. If we're really that afraid of non-magic magic numbers I'd maybe accept a `MUG_BYTES` constant somewhere... But even then that's somewhat fake because we still manually index `0` through `3` in various places, and other such mug-size-based logic.

Reply:
After a second review, I realized that `mex_l` most likely should be `mug_l`, in which case the change that I suggested would result in `if ( dat_i < sizeof(mug_l) ) {`, which maps very nicely into "if `dat_i` is smaller than the size of a mug". I see no problem with this.

In practice at this point in time in this version of the Urbit runtime, I understand that a mug is always four bytes. However, I'd argue a mug *happens* to be 4 bytes but is not defined to be so. In the hypothetical case in which we choose a different hash function or even a different variant of the same `murmur3` hash function, the mug--defined in this case to be a hash of a noun--is not necessarily four bytes. So while using `sizeof()` or a named constant `MUG_BYTES` does not address the use of literal indexes with respect to mugs elsewhere in the codebase, I think it better conveys in code the notion of what a mug is.